content/backend/graphql-node-typescript/8-authentication.md

```diff
@@ -0,0 +1,522 @@
+    where: { email: args.email },
+  });
+  if (!user) {
+    throw new Error("No such user found");
+  }
+
+  // 2
+  const valid = await compare(args.password, user.password);
+  if (!valid) {
+    throw new Error("Invalid password");
+  }
+
+  const token = sign({ userId: user.id }, APP_SECRET);
+
+  // 3
+  return {
+    token,
+    user,
+  };
+},
+```
+
+</Instruction>
+
+And if you'll open Playground, you should be able to login with the user you previously created:
+
+```graphql
+mutation {
+  login(email: "test@mail.com", password: "123456") {
+    token
+    user {
+      id
+      name
+      email
+    }
+  }
+}
+```
+
+> You should be able to get the information of the user. Please save the authentication token you get, we'll next that on the next step!
+
+Now on the `login` mutation!
+
+1. Instead of _creating_ a new `User` object, you're now using your `PrismaClient` instance to retrieve an existing `User` record by the `email` address that was sent along as an
+   argument in the `login` mutation. If no `User` with that email address was found, you're returning a corresponding error.
+1. The next step is to compare the provided password with the one that is stored in the database. If the two don't match, you're returning an error as well.
+1. In the end, you're returning `token` and `user` again.
+
+### Detecting the current user
+
+Now, we have our users database ready to use, and our next step is to be able to detect who's the current user that queries the server.
+
+To do that, we'll need to add the option to pass the authentication token along with our GraphQL operation.
+
+We are not going to use the GraphQL schema in this case, since we don't want to mix the authentication flow with the GraphQL contract that we have. So we'll use HTTP headers.
+
+We will pass the authentication token as a HTTP header, in the following form:
+
+```
+Authorization: "Bearer MY_TOKEN_HERE"
+```
+
+To add support for this kind of authentication in our server, we'll need to be able to access the raw incoming HTTP request, then verify the token and identify the current user. We also want to be able to tell who's the current authenticated user during our resolvers, so we'll inject the current user ID into the GraphQL `context`.
+
+So let's do that: 
+
+<Instruction>
+
+Start by adding the following imports to `src/index.ts`:
+
+```ts
+import { APP_SECRET, schema } from "./schema";
+import { JwtPayload, verify } from "jsonwebtoken";
+```
+
+Then, let's modify the `contextFactory` in `src/index.ts` function to do this flow:
+
+```ts
+contextFactory: () => {
+  let currentUser = null;
+
+  if (req.headers.authorization) {
+    // 1
+    const token = req.headers.authorization.split(" ")[1];
+    // 2
+    const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
+    // 3
+    const userId = tokenPayload.userId;
+    // 4
+    currentUser = prisma.user.findUnique({ where: { id: userId } });
+  }
```

Comment:
Right now if token is invalid or empty request doesn't go through and user cannot access login mutation. 
```suggestion
  if (req.headers.authorization) {
    // 1
    const token = req.headers.authorization.split(" ")[1];
    try {
      // 2
      const tokenPayload = verify(token, APP_SECRET) as JwtPayload;
      // 3
      const userId = tokenPayload.userId;
      // 4
      currentUser = prisma.user.findUnique({ where: { id: userId } });
    } catch {
      // We allow the request to go through so user can log in or access other unauthenticated operations
    }
  }
```